kroxylicious-filters/kroxylicious-record-validation/src/main/java/io/kroxylicious/proxy/filter/schema/validation/bytebuf/JsonSchemaBytebufValidator.java

```diff
@@ -12,31 +12,89 @@
+                true, buildHeaderHandler(true),
+                false, buildHeaderHandler(false));
+        this.idHandler = Map.of(
+                true, buildIdHandler(true),
+                false, buildIdHandler(false));
     }
 
     @Override
-    public CompletionStage<Result> validate(ByteBuffer buffer, int length, Record record, boolean isKey) {
+    public CompletionStage<Result> validate(ByteBuffer buffer, int length, Record kafkaRecord, boolean isKey) {
         try {
-            JsonValidationResult jsonValidationResult = jsonValidator.validateByArtifactReference(buffer.clear());
+            // If the record includes a schema id, validate that it is consistent with the expected globalId.
+            Optional<Long> extractedGlobalId = extractGlobalIdFromRecord(buffer, length, kafkaRecord, isKey);
+            if (extractedGlobalId.filter(e -> !e.equals(globalId)).isPresent()) {
+                return CompletableFuture
+                        .completedStage(new Result(false, "Unexpected schema id in record (%d), expecting %d".formatted(extractedGlobalId.get(), globalId)));
+            }
+
+            JsonValidationResult jsonValidationResult = jsonValidator.validateByArtifactReference(buffer);
             return jsonValidationResult.success() ? Result.VALID
-                    : CompletableFuture.completedFuture(new Result(false, jsonValidationResult.toString()));
+                    : CompletableFuture.completedStage(new Result(false, jsonValidationResult.toString()));
         }
         catch (RuntimeException ex) {
-            return CompletableFuture.completedFuture(new Result(false, ex.getMessage()));
+            return CompletableFuture.completedStage(new Result(false, ex.getMessage()));
         }
     }
+
+    private Optional<Long> extractGlobalIdFromRecord(ByteBuffer buffer, int length, Record kafkaRecord, boolean isKey) {
+        if (kafkaRecord.headers().length > 0) {
+            var recordHeaders = new RecordHeaders(kafkaRecord.headers());
+            var ref = headerHandler.get(isKey).readHeaders(recordHeaders);
+
+            if (ref.getGlobalId() != null) {
+                return Optional.of(ref.getGlobalId());
+            }
+        }
+
+        var minBytes = 1 + idHandler.get(isKey).idSize();
+        if (buffer.remaining() > minBytes && buffer.get(buffer.position()) == AbstractKafkaSerDe.MAGIC_BYTE) {
+            buffer.get(); // ignore magic
+            return Optional.of(idHandler.get(isKey).readId(buffer).getGlobalId());
+        }
+        return Optional.empty();
+    }
+
+    @NonNull
+    private static DefaultHeadersHandler buildHeaderHandler(boolean isKey) {
+        var handler = new DefaultHeadersHandler();
+        handler.configure(Map.of(), isKey);
+        return handler;
+    }
+
+    @NonNull
+    private static DefaultIdHandler buildIdHandler(boolean isKey) {
+        var handler = new DefaultIdHandler();
+        handler.configure(Map.of(), isKey);
```

Comment:
we probably ought to accept configuration.

Reply:
Yep I think both bits of configurability can wait. The user is configuring a global id in filter config, so there's no need to tell the DefaultIdHandler whether it's global or content id.

The other thing is DefaultIdHandler is for 8 byte ids, for confluent compatibility the user could choose to write 4 byte ids with the Legacy4ByteIdHandler. This is likely to be used in conjunction with content ids so limited chance of being useful with the global ids.

Reply:
> for confluent compatibility the user could choose to write 4 byte ids with the Legacy4ByteIdHandler. 

There's currently no way to adjust the rest endpoint beyond host/port, so I doubt things will actually work with a confluent backend.

Reply:
Ah, I'm not thinking of actual confluent compatibility, just that the apicurio producer can choose to write embedded ids using 4 or 8 bytes. Maybe some power users would use this to save bytes even without confluent schema registry in the mix.

See `apicurio.registry.id-handler` https://www.apicur.io/registry/docs/apicurio-registry/2.6.x/getting-started/assembly-configuring-kafka-client-serdes.html

Reply:
> Maybe some power users would use this to save bytes even without confluent schema registry in the mix.

Indeed.   It was that that originally got me thinking about all this in the first place.  The fact that the validators would choke if they encountered a record with the magic byte at the start of the value.

I have no sense how common this use-case might be.